chainntnfs/txnotifier_test.go

```diff
@@ -413,16 +417,498 @@ func TestTxConfHistoricalDispatch(t *testing.T) {
+	confNtfns := make([]*chainntnfs.ConfNtfn, numNtfns)
+	for i := uint64(0); i < numNtfns; i++ {
+		confNtfns[i] = &chainntnfs.ConfNtfn{
+			ConfID: i,
+			TxID:   &zeroHash,
+			Event:  chainntnfs.NewConfirmationEvent(1),
+		}
+		if _, err := n.RegisterConf(confNtfns[i]); err != nil {
+			t.Fatalf("unable to register conf ntfn #%d: %v", i, err)
+		}
+	}
+
+	// Ensure none of them have received the confirmation details.
+	for i, ntfn := range confNtfns {
+		select {
+		case <-ntfn.Event.Confirmed:
+			t.Fatalf("request #%d received unexpected confirmation "+
+				"notification", i)
+		default:
+		}
+	}
+
+	// We'll assume a historical rescan was dispatched and found the
+	// following confirmation details. We'll let the notifier know so that
+	// it can stop watching at tip.
+	expectedConfDetails := &chainntnfs.TxConfirmation{
+		BlockHeight: startingHeight - 1,
+	}
+	err := n.UpdateConfDetails(*confNtfns[0].TxID, expectedConfDetails)
+	if err != nil {
+		t.Fatalf("unable to update conf details: %v", err)
+	}
+
+	// With the confirmation details retrieved, each client should now have
+	// been notified of the confirmation.
+	for i, ntfn := range confNtfns {
+		select {
+		case confDetails := <-ntfn.Event.Confirmed:
+			assertConfDetails(t, confDetails, expectedConfDetails)
+		default:
+			t.Fatalf("request #%d expected to received "+
+				"confirmation notification", i)
+		}
+	}
+
+	// Similarly, we'll do the same thing but for spend notifications.
+	spendNtfns := make([]*chainntnfs.SpendNtfn, numNtfns)
+	for i := uint64(0); i < numNtfns; i++ {
+		spendNtfns[i] = &chainntnfs.SpendNtfn{
+			SpendID:  i,
+			OutPoint: zeroOutPoint,
+			Event:    chainntnfs.NewSpendEvent(nil),
+		}
+		if _, err := n.RegisterSpend(spendNtfns[i]); err != nil {
+			t.Fatalf("unable to register spend ntfn #%d: %v", i, err)
+		}
+	}
+
+	// Ensure none of them have received the spend details.
+	for i, ntfn := range spendNtfns {
+		select {
+		case <-ntfn.Event.Spend:
+			t.Fatalf("request #%d received unexpected spend "+
+				"notification", i)
+		default:
+		}
+	}
+
+	// We'll assume a historical rescan was dispatched and found the
+	// following spend details. We'll let the notifier know so that it can
+	// stop watching at tip.
+	expectedSpendDetails := &chainntnfs.SpendDetail{
+		SpentOutPoint:     &spendNtfns[0].OutPoint,
+		SpenderTxHash:     &zeroHash,
+		SpendingTx:        wire.NewMsgTx(2),
+		SpenderInputIndex: 0,
+		SpendingHeight:    startingHeight - 1,
+	}
+	err = n.UpdateSpendDetails(spendNtfns[0].OutPoint, expectedSpendDetails)
+	if err != nil {
+		t.Fatalf("unable to update spend details: %v", err)
+	}
+
+	// With the spend details retrieved, each client should now have been
+	// notified of the spend.
+	for i, ntfn := range spendNtfns {
+		select {
+		case spendDetails := <-ntfn.Event.Spend:
+			assertSpendDetails(t, spendDetails, expectedSpendDetails)
+		default:
+			t.Fatalf("request #%d expected to received spend "+
+				"notification", i)
+		}
+	}
```

Comment:
Good test! 💯 

Attempt to register a new client after update spend details was called, that we expect to get notified?